script.module.clouddrive.common/clouddrive/common/service/export.py

```diff
@@ -0,0 +1,245 @@
+        self.cleanup_export_map()
+        monitor = KodiUtils.get_system_monitor()
+        startup = True
+        while not self.abort:
+            try:
+                now = datetime.datetime.now()
+                export_map = self.get_export_map()
+                if export_map:
+                    self.process_schedules(export_map, now, startup)
+                    self.process_watch(export_map)
+            except Exception as e:
+                ErrorReport.handle_exception(e)
+            startup = False
+            if monitor.waitForAbort(60):
+                break
+        del monitor
+        del self.provider
+        Logger.notice('Service stopped.')
+    
+    def process_schedules(self, export_map, now, startup=False):
+        Logger.debug('now: %s, startup: %s' % (Utils.str(now), Utils.str(startup)))
+        export_list = []
+        if startup:
+            export_list.extend(Utils.get_safe_value(export_map, self._startup_type, []))
+        else:
+            at = '%02d:%02d' % (now.hour, now.minute,)
+            Logger.debug('at: %s' % Utils.str(at))
+            daily_list = Utils.get_safe_value(export_map, Utils.str(ExportScheduleDialog._daily_type) + at, [])
+            export_list.extend(daily_list)
+            Logger.debug('daily_list: %s' % Utils.str(daily_list))
+            weekday = now.weekday() + 11
+            weekday_list = Utils.get_safe_value(export_map, Utils.str(weekday) + at, [])
+            export_list.extend(weekday_list)
+            Logger.debug('weekday_list: %s' % Utils.str(weekday_list))
+        Logger.debug('export_list: %s' % Utils.str(export_list) )
+        for export in export_list:
+            self.run_export(export)
+    
+    def run_export(self, export):
+        export['exporting'] = True
+        params = {'action':'_run_export', 'content_type': Utils.get_safe_value(export, 'content_type', ''), 'driveid': Utils.get_safe_value(export, 'driveid', ''), 'item_id': export['id']}
+        KodiUtils.run_plugin(self.addonid, params, False)
+    
+    def process_watch(self, export_map):
+        exports = Utils.get_safe_value(export_map, 'watch', [])
+        update_library = {}
+        changes_by_drive = {}
+        for export in exports:
+            item_id = export['id']
+            driveid = export['driveid']
+            if driveid in changes_by_drive:
+                changes = changes_by_drive[driveid]
+            else:
+                self.provider.configure(self._account_manager, export['driveid'])
+                changes = self.provider.changes()
+                changes_by_drive[driveid] = changes
+            items_info = self.export_manager.get_items_info(item_id)
+            if items_info:
+                if changes and not Utils.get_safe_value(export, 'exporting', False):
+                    Logger.debug('*** Processing changes for export "%s" in %s' % (export['name'], export['destination_folder']))
+                    while True:
```

Comment:
this is a forever blocking loop, use monitor to listen for requested aborts

Reply:
Working on it

Reply:
Just reviewing this 'while True'. It's not a forever blocking loop. Python doesn't have a do-while loop (just in case, no offence: it's a while but you need the code executed at least one time). 
So the code could be read as:
do {
 #bla bla
} while not (changes_done and changes_retry)

I know the while True makes a lot of noise (it did for me when you mentioned it), but reviewing it makes me remember why I put it there.
If I put a monitor.onAbortRequested() there is like puting the abort check in all whiles and fors.
Let me know

Reply:
ok